### PR TITLE
Feature/runningwater

### PR DESCRIPTION
**Pull request by Romeeey**
# Checklist
* [x] Start the game
* [x] Check running water for rivers and waterfalls
* [ ] Check gothic skybox

## Scene
* [ ] FeatureFlags reverted to right value

# Testing
* [ ] Merged _main_ into this branch and tested with the latest features
* [ ] Tested with PCVR
* [ ] Tested with Pico / Quest

## Dependencies
* [ ] If libphoenix-shared.dll is changed, .so is also changed

### CODE REVIEW

**Comment by JaXt0r** [comment]
Tested:
* Water flows - but feels very fast. e.g. river in front of OC flows as fast as waterfalls. (Possible improvement?)
* Is this the skybox you meant? If so, then ok.
image

**Comment by JaXt0r** [comment]
If I look at the texture at runtime, it seems skybox isn't used.
image

File: Assets/GothicVR/Resources/Materials/GothicSky.mat
Comment (JaXt0r) [inline Assets/GothicVR/Resources/Materials/GothicSky.mat:1]: Do we really need this? Frank already changed material usage in his other branch. Please align with him and sanitize.

File: Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs [hunk 1]
```diff
@@ -510,5 +521,29 @@ protected Material GetEmptyMaterial()
 
             return material;
         }
+
+        private Material GetWaterMaterial(PxMaterialData materialData)
```
Comment (JaXt0r) [inline Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs:525]: Well done. :-)

File: Assets/GothicVR/Scripts/UI/TextureManager.cs [hunk 2]
```diff
@@ -53,9 +53,8 @@ public void LoadCustomTextures()
         arrowtexture = AssetCache.I.TryGetTexture("U.TGA");
         arrowmaterial.mainTexture = arrowtexture;
 
-        // FIXME - No material found in Assets. commenting it out right now.
-        // skytexture = AssetCache.I.TryGetTexture("SKYDAY_LAYER0_A0-C.TEX");
-        // skymaterial.mainTexture = skytexture;
+        skytexture = AssetCache.I.TryGetTexture("SKYDAY_LAYER0_A0-C.TEX");
```
Comment (JaXt0r) [inline Assets/GothicVR/Scripts/UI/TextureManager.cs:56]: PS: You don't need to add -C.TEX or .TGA. The TryGetTexture() method is removing the suffix and will test it on it's own.
No bug, just something I wanted to mention. :-)

File: ProjectSettings/GraphicsSettings.asset [hunk 3]
```diff
@@ -29,6 +29,7 @@ GraphicsSettings:
   m_AlwaysIncludedShaders:
   - {fileID: 4800000, guid: 650dd9526735d5b46b79224bc6e94025, type: 3}
   - {fileID: -6465566751694194690, guid: 9b4e681081e2b4c469111bb649e2f7ee, type: 3}
+  - {fileID: -6465566751694194690, guid: 6a56f4d49ccc01c4caeafc424f9766ed, type: 3}
```
Comment (JaXt0r) [inline ProjectSettings/GraphicsSettings.asset:32]: What's this? The skybox? If wanted, then fine. :-)

**Review by JaXt0r — CHANGES_REQUESTED** [review]
A request for optimization of material usage and I think the skybox isn't shown.

File: ProjectSettings/GraphicsSettings.asset (hunk 3 above)
Comment (Romeeey) [reply ProjectSettings/GraphicsSettings.asset:32]: This is vinces texture scrolling shader, that is used for the running water.

File: Assets/GothicVR/Resources/Materials/GothicSky.mat
Comment (Romeeey) [reply Assets/GothicVR/Resources/Materials/GothicSky.mat:1]: I'll wait for his implementation and then change it.

File: Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs (hunk 1 above)
Comment (Romeeey) [reply Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs:525]: That was your implementation =D
I simply modified it for the abstractmeshcreator.

**Comment by Romeeey** [comment]
> Tested:
> 
> * Water flows - but feels very fast. e.g. river in front of OC flows as fast as waterfalls. (Possible improvement?)

There is certainly space for improvement.

> * Is this the skybox you meant? If so, then ok.
>   image

Let's discuss this offline, before deciding on next steps.

File: Assets/GothicVR/Scripts/UI/TextureManager.cs (hunk 2 above)
Comment (Romeeey) [reply Assets/GothicVR/Scripts/UI/TextureManager.cs:56]: Ok. Cool. I will change it.

File: Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs [hunk 4]
```diff
@@ -487,5 +499,29 @@ protected Material GetEmptyMaterial()
...
+        {
+            var shader = Shader.Find(waterShader);
+            Material material = new Material(shader);
+
+            // FIXME - Here we need to see if we can change materialData.waveSpeed
+
+            switch (materialData.name)
```
Comment (JaXt0r) [inline Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs:510]: Please don't! This is hard coded and will prevent mods etc. let's properly check how to do.
1. Check on OpenGothic and ask Try/Luis
2. Let's define a default speed for now. Maybe just something slower.

**Comment by Romeeey** [comment]
> > Tested:
> > 
> > * Water flows - but feels very fast. e.g. river in front of OC flows as fast as waterfalls. (Possible improvement?)
> 
> There is certainly space for improvement.
> 
> > * Is this the skybox you meant? If so, then ok.
> >   image
> 
> Let's discuss this offline, before deciding on next steps.

Merged from main
Skybox fixed added again
Improved shader for flowing water 
Changed running water speeds and directions manually (is a temporary solution until the proper water implementation is done)

Ready to merge.

File: Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs (hunk 4 above)
Comment (JaXt0r) [inline Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs:510]: You could give this a try. This is the Vector, OpenGothic is using for animation. Sqrt from Vector is definitely not the intended idea, but see for yourself: waterfallls look fine.
```suggestion
            material.SetFloat("_ScrollSpeed", -900000 * materialData.animMapDir.ToUnityVector().SqrMagnitude());
```

File: Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs [hunk 5]
```diff
@@ -13,6 +13,7 @@
 using PxCs.Data.Struct;
 using PxCs.Data.Vob;
 using PxCs.Interface;
+using Unity.VisualScripting;
```
Comment (JaXt0r) [inline Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs:16]: not needed.

**Review by JaXt0r — CHANGES_REQUESTED** [review]

File: Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs (hunk 5 above)
Comment (Romeeey) [reply Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs]: Removed.

File: Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs (hunk 4 above)
Comment (Romeeey) [reply Assets/GothicVR/Scripts/Creator/Meshes/AbstractMeshCreator.cs:507]: That's a cool approach using proper material data. Nevertheless it still causes some problems in the swamp / water transition zone and the rivers. I would go for the hardcoded ones for now and go for a proper implementation in the near future.

**Pull request merged**